idl2library.cmake

```
@@ -192,19 +193,52 @@ function(idl2library)
         # Group the IDL projects together
         set_target_properties(${current_idl_target} PROPERTIES FOLDER IDL)
         target_compile_definitions(${current_idl_target} PUBLIC _HAS_AUTO_PTR_ETC=1 _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING)
+            
+        #Set the PUBLIC_HEADER for the target
+        if (${idl2library_INSTALL_OUTPUT})
+            get_target_property(HEADER_DIR ${current_idl_target} OPENDDS_GENERATED_DIRECTORY)
+            string(REPLACE "_idl" "" baseName "${HEADER_DIR}/${current_idl_target}")
+            
+            get_property(all_public_headers TARGET ${current_idl_target} PROPERTY PUBLIC_HEADER)
+            
+            list(APPEND all_public_headers "${baseName}TypeSupportImpl.h")
+            list(APPEND all_public_headers "${baseName}TypeSupportC.h")
+            list(APPEND all_public_headers "${baseName}TypeSupportC.inl")
+            list(APPEND all_public_headers "${baseName}TypeSupportS.h")
+            list(APPEND all_public_headers "${baseName}C.h")
+            list(APPEND all_public_headers "${baseName}S.h")
+            list(APPEND all_public_headers "${baseName}C.inl")
+            
+            set_property(TARGET ${current_idl_target} PROPERTY PUBLIC_HEADER ${all_public_headers})
+            
+            target_include_directories(${current_idl_target} PUBLIC
+                 $<INSTALL_INTERFACE:include>
+            )
```

OpenDDS has a  [function for installing generated files](https://opendds.readthedocs.io/en/latest-release/devguide/building/cmake.html#cmake-func-opendds_install_interface_files).

---

Would that take care of setting the public headers?  How would you suggest I use the function in idl2dll for installing files?

---

It doesn't use public headers, [it uses `install`](https://github.com/OpenDDS/OpenDDS/blob/cc443c3ba7a792a8182a96c8ebb3243f1756e0df/cmake/opendds_utils.cmake#L255) directly, which is another way to do basically the same thing. If you really need to use `PUBLIC_HEADER`, you could get [`OPENDDS_ALL_INTERFACE_FILES`](https://opendds.readthedocs.io/en/latest-release/devguide/building/cmake.html#cmake-prop-opendds_target_sources-OPENDDS_ALL_INTERFACE_FILES), and they could transformed into a list for `PUBLIC_HEADER`.

For `opendds_install_interface_files`, I think it's just:

```suggestion
            opendds_install_interface_files(${current_idl_target})
```

Since it looks like it's not putting it in a directory inside `include`. For it to work though, the IDL file in `opendds_target_sources` needs to be marked as `PUBLIC`:

https://github.com/OpenDDS/OpenDDW/blob/00b0a6cec249b04146635da62bc70b7c5e39f174/idl2library.cmake#L182-L183